### PR TITLE
Attempt of writing a GPU kernel to address #2737

### PR DESCRIPTION
**Pull request by tztsai**
I wrote a GPU kernel for the function `compute_matrix_for_linear_operation` in [matrix_solver_utils.jl](src/Solvers/matrix_solver_utils.jl), attempting to parallel `linear_operation!` that produces the matrix `A`. Since in parallelization, values in `rowval` and `nzval` are appended in undetermined order, they must be sorted after the GPU computation to produce the correct inputs to construct the CUDA sparse matrix.

### CODE REVIEW

File: src/Solvers/matrix_solver_utils.jl [hunk 1]
```diff
@@ -154,6 +154,40 @@ function compute_matrix_for_linear_operation(::CPU, template_field, linear_opera
     return A
 end
 
+function compute_matrix_for_linear_operation_kernel!(Aeᵢⱼₖ, eᵢⱼₖ, linear_operation!, colptr, rowval, nzval, args...)
+    i, j, k = (blockIdx().x - 1) * blockDim().x + threadIdx().x,
+              (blockIdx().y - 1) * blockDim().y + threadIdx().y,
+              (blockIdx().z - 1) * blockDim().z + threadIdx().z
```
Comment (glwagner) [inline src/Solvers/matrix_solver_utils.jl:160]: is it possible to use `KernelAbstractions`?

**Comment by glwagner** [comment]
@simone-silvestri

**Comment by glwagner** [comment]
@navidcy

File: src/Solvers/matrix_solver_utils.jl (hunk 1 above)
Comment (TZTsai) [reply src/Solvers/matrix_solver_utils.jl]: I have updated the code to use `KernelAbstractions`.

**Comment by navidcy** [comment]
@TZTsai thanks for this! I'm wondering whether this worths your time and effort since after we [removed](https://github.com/CliMA/Oceananigans.jl/pull/3052) the multigrid pressure solver from the code perhaps this is not so much needed.

**Comment by glwagner** [comment]
@navidcy can't we move this PR to another project? You should also close the issue if it's not relevant.

**Comment by glwagner** [comment]
I thikn you could also migrate the PR to the other repo

**Comment by glwagner** [comment]
hm

**Comment by TZTsai** [comment]
> @TZTsai thanks for this! I'm wondering whether this worths your time and effort since after we [removed](https://github.com/CliMA/Oceananigans.jl/pull/3052) the multigrid pressure solver from the code perhaps this is not so much needed.

Hmm alright, but is `compute_matrix_for_linear_operation` still used in this project? If so it still provided some optimization.

File: src/Solvers/matrix_solver_utils.jl [hunk 2]
```diff
@@ -154,12 +154,39 @@ function compute_matrix_for_linear_operation(::CPU, template_field, linear_opera
     return A
 end
 
+@kernel function compute_matrix_for_linear_operation_kernel!(Aeᵢⱼₖ, eᵢⱼₖ, linear_operation!, indices, values, args...)
+    i, j, k = @index(Global, NTuple)
```
Comment (simone-silvestri) [inline src/Solvers/matrix_solver_utils.jl:158]: you are launching the kernel with a linear index:
```julia
kernel! = compute_matrix_for_linear_operation_kernel!(device(GPU()), min(256, N), N)
```
where `N` is a scalar, so there is only one index 
```julia
i = @index(Global, Linear)
```
if you want to have 3 indices you have to launch the kernel with a 3-tuple 
```julia
kernel! = compute_matrix_for_linear_operation_kernel!(device(GPU()), min(256, N), (Nx, Ny, Nz))
```

Comment (TZTsai) [reply src/Solvers/matrix_solver_utils.jl:158]: Thanks for pointing out! I have made the update now.

**Comment by glwagner** [comment]
> > @TZTsai thanks for this! I'm wondering whether this worths your time and effort since after we [removed](https://github.com/CliMA/Oceananigans.jl/pull/3052) the multigrid pressure solver from the code perhaps this is not so much needed.
> 
> Alright. But is `compute_matrix_for_linear_operation` still used in this project? If so it may still provide some optimization.

@navidcy can you answer @TZTsai's question?

**Comment by navidcy** [comment]
> > @TZTsai thanks for this! I'm wondering whether this worths your time and effort since after we [removed](https://github.com/CliMA/Oceananigans.jl/pull/3052) the multigrid pressure solver from the code perhaps this is not so much needed.
> 
> Alright. But is `compute_matrix_for_linear_operation` still used in this project? If so it may still provide some optimization.

Sorry I missed the question!

I think in principle it could be used for the construction of the Heptadiagonal solver matrix (cc @simone-silvestri) but even for that the matrix is actually hardcoded for the particular free-surface problem. Right @simone-silvestri?

**Comment by navidcy** [comment]
@TZTsai is this ready for review? I'm happy to review it.

Out of curiosity, have you checked that you reproduce the matrix for a given linear operation before and after the changes in the PR using GPU?

File: src/Solvers/matrix_solver_utils.jl [hunk 3]
```diff
@@ -154,12 +154,39 @@ function compute_matrix_for_linear_operation(::CPU, template_field, linear_opera
...
+    Nx, Ny, Nz = size(eᵢⱼₖ)
+    N = Nx * Ny * Nz
+
+    if i <= Nx && j <= Ny && k <= Nz
+        parent(eᵢⱼₖ)  .= 0
+        parent(Aeᵢⱼₖ) .= 0
+
+        @inbounds eᵢⱼₖ[i, j, k] = 1
+
+        fill_halo_regions!(eᵢⱼₖ)
```
Comment (simone-silvestri) [inline src/Solvers/matrix_solver_utils.jl:168]: unfortunately, it is not possible to call `fill_halo_regions!` inside a kernel. `fill_halo_regions!` is a function that _calls_ kernels and to call a kernel from a kernel you need dynamic parallelism (we do not make use of that in oceananigans). This has to do with the fact that a kernel is tied to a certain thread configuration, so calling a kernel from a kernel you might have to redefine the thread structure.

File: src/Solvers/matrix_solver_utils.jl [hunk 4]
```diff
@@ -154,12 +154,39 @@ function compute_matrix_for_linear_operation(::CPU, template_field, linear_opera
...
+
+        linear_operation!(Aeᵢⱼₖ, eᵢⱼₖ, args...)
+
+        n = Ny*Nx*(k-1) + Nx*(j-1) + i  # n-th column of A
+
+        for (m, v) in enumerate(Aeᵢⱼₖ)  # m-th row of A
+            if v != 0
+                append!(indices, (m-1)*N + (n-1))  # zero-based
```
Comment (simone-silvestri) [inline src/Solvers/matrix_solver_utils.jl:176]: Not sure `append!` works on a GPU, I can give it a try

**Comment by glwagner** [comment]
@simone-silvestri can you comment on where this code is / will be used?

File: src/Solvers/matrix_solver_utils.jl [hunk 5]
```diff
@@ -174,34 +201,15 @@ function compute_matrix_for_linear_operation(::GPU, template_field, linear_opera
...
-        end
-
-        CUDA.@allowscalar colptr[Ny*Nx*(k-1) + Nx*(j-1) + i + 1] = colptr[Ny*Nx*(k-1) + Nx*(j-1) + i] + count
-    end
+    rows = div.(indices, N) .+ 1
+    cols = mod.(indices, N) .+ 1
 
-    return CuSparseMatrixCSC(colptr, rowval, nzval, (Nx*Ny*Nz, Nx*Ny*Nz))
+    # Is it necessary to use the CSC format?
+    return CuSparseMatrixCSC(CuSparseMatrixCOO(rows, cols, values, (N, N)))
```
[the remark this answers is not in the document]
Comment (TZTsai) [reply src/Solvers/matrix_solver_utils.jl:214]: OK

**Pull request closed without merging**